### PR TITLE
Fix materialize + S3 not working.

### PR DESCRIPTION
**Pull request by eric-anderson**
Existing code was relying on python pathlib to do most operations rather than running them through the pyarrow filesystem.  Fix that, and add a test with a fake pyarrow filesystem to make sure we run with a path that does not exist in the local filesystem.

### CODE REVIEW

File: lib/sycamore/sycamore/tests/unit/inmempyarrowfs.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+            if not k.startswith(dir):
+                continue
+            if "/" in k[dlen:]:
+                continue
+            ret.append(FileInfo(str(k), type=FileType.File))
+
+        return ret
+
+    def move(self, src, dest):
+        assert False, "unimplemented"
```
Comment (HenryL27) [inline lib/sycamore/sycamore/tests/unit/inmempyarrowfs.py:65]: there's also `raise NotImplementedError()`. Not sure what you prefer.

File: lib/sycamore/sycamore/tests/unit/test_materialize.py [hunk 2]
```diff
@@ -121,6 +124,7 @@ def doc_to_name3(doc):
             self.check_files(tmpdir, ext=".test3")
 
             files = glob.glob(tmpdir + "/*")
+            logging.error(f"ERIC {files}")
```
Comment (HenryL27) [inline lib/sycamore/sycamore/tests/unit/test_materialize.py:127]: cleaning

File: lib/sycamore/sycamore/tests/unit/test_materialize.py [hunk 3]
```diff
@@ -250,24 +253,26 @@ def doc_to_name4(doc):
...
             a._path["clean"] = True
             ds.execute()
             files = [f for f in Path(tmpdir).rglob("*")]
-            logging.info(f"TO Found-3 {files}")
             assert len([f for f in files if ".test4" in str(f)]) == 3 + 1 + 3 + 2
 
 
+def any_id(d):
+    if isinstance(d, MetadataDocument):
+        return str(d.metadata)
```
Comment (HenryL27) [inline lib/sycamore/sycamore/tests/unit/test_materialize.py:267]: this seems weird? but ok

File: lib/sycamore/sycamore/tests/unit/test_materialize.py [hunk 4]
```diff
@@ -331,3 +336,24 @@ def test_materialize_read(self):
             Path(tmpdir).rmdir()
             with pytest.raises(ValueError):
                 ds.take_all()
+
+
+class TestAllViaPyarrowFS(unittest.TestCase):
```
Comment (HenryL27) [inline lib/sycamore/sycamore/tests/unit/test_materialize.py:341]: Is it worth adding a ray test in here?

**Review by HenryL27 — APPROVED** [review]

File: lib/sycamore/sycamore/tests/unit/inmempyarrowfs.py (hunk 1 above)
Comment (eric-anderson) [reply lib/sycamore/sycamore/tests/unit/inmempyarrowfs.py:65]: switched. I have no strong preference.

File: lib/sycamore/sycamore/tests/unit/test_materialize.py (hunk 2 above)
Comment (eric-anderson) [reply lib/sycamore/sycamore/tests/unit/test_materialize.py:127]: removed.

File: lib/sycamore/sycamore/tests/unit/test_materialize.py (hunk 3 above)
Comment (eric-anderson) [reply lib/sycamore/sycamore/tests/unit/test_materialize.py:267]: ack

File: lib/sycamore/sycamore/tests/unit/test_materialize.py (hunk 4 above)
Comment (eric-anderson) [reply lib/sycamore/sycamore/tests/unit/test_materialize.py:341]: Per discussion, not needed. There is an integration test that uses ray.

File: lib/sycamore/sycamore/materialize.py [hunk 5]
```diff
@@ -328,24 +338,29 @@ def before(node):
...
-        if self.directory is None:
+        directory = path.pop("root", None)
+        self._path = path
+
+        if directory is None:
             from datetime import datetime
             import tempfile
 
             now = datetime.now().replace(microsecond=0)
             dir = Path(tempfile.gettempdir()) / f"materialize.{now.isoformat()}"
```
Comment (HenryL27) [inline lib/sycamore/sycamore/materialize.py:360]: does this ever leak storage?

**Review by HenryL27 — APPROVED** [review]
I dont see anything obviously wrong here and I'll trust you that this fixes the automat

File: lib/sycamore/sycamore/materialize.py (hunk 5 above)
Comment (eric-anderson) [reply lib/sycamore/sycamore/materialize.py:360]: It's intended to "leak". You're asking it to materialize so it's trying to leave it around.

**Pull request merged**